Fix bugs with setup and teardown

DmitrySharabin opened the pull request:
- Maintain proper sequencing when hooks (`beforeEach`/`afterEach`, `beforeAll`/`afterAll`) are used
- Fall back to parallel execution (using `Promise.all`) when no hooks exist

Now, we might end up (I already was bit by this bug) calling `afterAll` before `afterEach` for some tests. In most cases, `beforeEach` for all tests is called before the first test runs. There are probably some more issues we haven't encountered yet. This PR should fix those bugs.

LeaVerou reviewed (commented):
The way I see it, we can still do parallel execution with these, as long as:
- For each test, beforeEach is always executed before run and afterEach after run 
- beforeAll and afterAll are executed before and after all tests (regardless of their order)

DmitrySharabin commented:
> The way I see it, we can still do parallel execution with these, as long as:
> 
> * For each test, beforeEach is always executed before run and afterEach after run
> * beforeAll and afterAll are executed before and after all tests (regardless of their order)

You are absolutely right. Previously, I was confused seeing `beforeEach` of a test before `run` of the previous one. But now I understand it's OK if the order of hooks for individual tests is preserved. It should be fixed now.

LeaVerou commented on src/classes/TestResult.js:
```diff
@@ -142,24 +142,28 @@ export default class TestResult extends BubblingEventTarget {
 				if (this.test.isTest) {
 					if (this.test.skip) {
-						this.skip();
+						return this.skip();
```
If we're skipping a test, its before/after hooks should also be skipped.

DmitrySharabin replied:
Done

LeaVerou reviewed (changes requested):
This looks a bit too complicated and I have some reservations about whether the complexity is necessary, but I may be missing some of the nuances of this part of the code.
- `allFinished` is added dynamically, which means it could not be present when awaited from another part of the code
- if `runAll()` resolves when all tests have ran, why do we need `allFinished`? Can't we just `Promise.allSettled()` its result?
- Remember that in theory, tests can be run in isolation, without running the whole group. At first glance it seems to me that this wouldn't run beforeAll/afterAll, but I didn't look too closely.
-

DmitrySharabin commented:
> This looks a bit too complicated and I have some reservations about whether the complexity is necessary, but I may be missing some of the nuances of this part of the code.
> 
> * `allFinished` is added dynamically, which means it could not be present when awaited from another part of the code
> * if `runAll()` resolves when all tests have ran, why do we need `allFinished`? Can't we just `Promise.allSettled()` its result?
> * Remember that in theory, tests can be run in isolation, without running the whole group. At first glance it seems to me that this wouldn't run beforeAll/afterAll, but I didn't look too closely.

What an excellent feedback (as always)! Thank you.

I tried to address it:
- moved `beforeEach`/`afterEach` to `TestResult.run`, so if a test is skipped, those hooks are not called
- simplified the code (got rid of redundant `allFinished` in favor of `Promise.allSettled`)
- `beforeAll`/`afterAll` should now run when a test is running in isolation

I'll be grateful if you could find some time to review this PR one more time.
